src/dos/dos_keyboard_layout.cpp

```diff
@@ -958,306 +1001,268 @@ KeyboardErrorCode KeyboardLayout::ReadCodePageFile(const char *requested_cp_file
+        const auto old_font_origin      = loaded_layout ? loaded_layout->GetCodePageFontOrigin() :
+                                                          CodePageFontOrigin::Unknown;
+        const auto old_keyboard_layout  = DOS_GetLoadedLayout();
+
+        if (code_page_autodetect && !cpi_file.empty()) {
+                assert(false); // Incorrect arguments
+                return KeyboardLayoutResult::NoCodePageInCpiFile;
+        }
+
+        auto new_layout = std::make_unique<KeyboardLayout>();
+
+        std::vector<uint16_t> code_pages = {};
+
+        if (!IS_EGAVGA_ARCH) {
+                // Can't set custom code page on pre-EGA display adapters
+                if (code_page_autodetect) {
+                        code_pages.push_back(DefaultCodePage);
+                } else if (code_page != DefaultCodePage || !cpi_file.empty()) {
+                        return KeyboardLayoutResult::IncompatibleMachine;
+                }
+        } else if (code_page_autodetect) {
+                // Get the preferred code page
+                const auto bundled = DOS_GetBundledCodePage(keyboard_layout);
+                code_pages = DOS_SuggestBetterCodePages(bundled, keyboard_layout);
+                code_pages.push_back(bundled);
+        } else {
+                code_pages.push_back(code_page);
+        }
+
+	// Try to read the keyboard layout
+	KeyboardLayoutResult result = KeyboardLayoutResult::LayoutNotKnown;
+        for (const auto try_code_page : code_pages) {
+                code_page = try_code_page;
+                result = new_layout->ReadKeyboardFile(keyboard_layout, code_page);
+                if (result == KeyboardLayoutResult::OK) {
+                        break;
+                }
+        }
+
+        // Allow using 'us' layout with any code page
+        if ((result == KeyboardLayoutResult::LayoutNotKnown ||
+             result == KeyboardLayoutResult::NoLayoutForCodePage) &&
+            iequals(keyboard_layout, "US")) {
+                // Allow using 'us' layout with any code page
+                result = new_layout->ReadKeyboardFile(keyboard_layout, DefaultCodePage);
+        }
+
+        // If no suitable code page found, terminate
+        if (result != KeyboardLayoutResult::OK) {
+                if (result == KeyboardLayoutResult::NoLayoutForCodePage &&
+                    code_page_autodetect) {
+                        return KeyboardLayoutResult::IncompatibleMachine;
+                }
+                return result;
+        }
+
+	// Keyboard layout loaded successfully, change code page accordingly
+        if (!IS_EGAVGA_ARCH || (prefer_rom_font && code_page == DefaultCodePage)) {
+                new_layout->SetRomFont();
+        } else {
+                result = new_layout->ReadCodePageFile(cpi_file, code_page);
+                if (result != KeyboardLayoutResult::OK) {
+                        return result;
+                }
+        }
+
```

Comment:
Noticed that comment is using tab indenting but others are using spaces indenting. Might have to reformat the whole file?

Reply:
Should be fixed now.